File: modules/coupon-infrastructure/src/main/java/com/thanh0x/coursedeal/utils/UrlValidator.java
package com.thanh0x.coursedeal.utils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.InetAddress;
import java.net.URI;
import java.net.UnknownHostException;
import java.util.Set;

/**
 * Utility class to prevent Server-Side Request Forgery (SSRF) attacks.
 * It ensures that the application only makes requests to allowed domains and blocks internal IP ranges.
 */
public class UrlValidator {
    private static final Logger log = LoggerFactory.getLogger(UrlValidator.class);

    private static final Set<String> ALLOWED_DOMAINS = Set.of(
            "udemy.com",
            "www.udemy.com",
            "jobs.e-next.in",
            "cdn.real.discount"
    );

    /**
     * Validates a URL to prevent SSRF.
     *
     * @param urlString The URL to validate.
     * @return true if the URL is safe, false otherwise.
     */
    public static boolean isSafeUrl(String urlString) {
        try {
            URI uri = URI.create(urlString);
            String host = uri.getHost();

            if (host == null || host.isBlank()) {
                log.warn("SSRF Blocked: URL has no host: {}", urlString);
                return false;
            }

            // 1. Check Allowlist
            boolean isAllowed = ALLOWED_DOMAINS.stream()
                    .anyMatch(domain -> host.equals(domain) || host.endsWith("." + domain));

            if (!isAllowed) {
                log.warn("SSRF Blocked: Domain not in allowlist: {}", host);
                return false;
            }

            // 2. Check for Private IP Ranges
            InetAddress address = InetAddress.getByName(host);
            if (address.isLoopbackAddress() || address.isSiteLocalAddress() || address.isLinkLocalAddress()) {
                log.warn("SSRF Blocked: Attempt to access internal/private IP: {} ({})", host, address.getHostAddress());
                return false;
            }

            return true;
        } catch (UnknownHostException e) {
            log.warn("SSRF Blocked: Unable to resolve host: {}", urlString);
            return false;
        } catch (Exception e) {
            log.warn("SSRF Blocked: Invalid URL format: {}", urlString);
            return false;
        }
    }
}

File: modules/coupon-infrastructure/src/main/kotlin/com/thanh0x/coursedeal/utils/UrlValidator.kt
package com.thanh0x.coursedeal.utils

import com.thanh0x.coursedeal.config.logger
import java.net.InetAddress
import java.net.URI
import java.net.UnknownHostException

/**
 * Utility object to prevent Server-Side Request Forgery (SSRF) attacks.
 */
object UrlValidator {
    private val log = logger()

    private val ALLOWED_DOMAINS =
        setOf(
            "udemy.com",
            "www.udemy.com",
            "jobs.e-next.in",
            "cdn.real.discount",
        )

    /**
     * Validates a URL to prevent SSRF.
     */
    @JvmStatic
    fun isSafeUrl(urlString: String?): Boolean {
        return try {
            if (urlString.isNullOrBlank()) return false

            val uri = URI.create(urlString)
            val host = uri.host

            if (host.isNullOrBlank()) {
                return false
            }

            // 1. Check Allowlist
            val isAllowed =
                ALLOWED_DOMAINS.any { domain ->
                    host.equals(domain, ignoreCase = true) || host.lowercase().endsWith(".$domain")
                }

            if (isAllowed) {
                return true
            }

            // 2. Block direct IP access or non-whitelisted domains
            if (isInternalIP(host)) {
                log.warn("SSRF Blocked: Attempt to access internal/private IP: {}", host)
                return false
            }

            log.warn("SSRF Blocked: Domain not in allowlist: {}", host)
            false
        } catch (e: Exception) {
            false
        }
    }

    private fun isInternalIP(host: String): Boolean {
        return try {
            // If it's not a whitelisted domain, we check if it resolves to a private IP
            val address = InetAddress.getByName(host)
            address.isLoopbackAddress || address.isSiteLocalAddress || address.isLinkLocalAddress
        } catch (e: UnknownHostException) {
            // If it can't be resolved and isn't whitelisted, it's unsafe
            true
        }
    }
}
